fuse3: 3.11.0 -> 3.16.2

6t8k opened the pull request:
## Description of changes

https://github.com/libfuse/libfuse/blob/fuse-3.16.2/ChangeLog.rst#libfuse-3162-2023-10-10
https://github.com/libfuse/libfuse/compare/fuse-3.11.0...fuse-3.16.2

---

One change can be expected to break some setups:
- Unsupported mount options are no longer silently accepted [[commit](https://github.com/libfuse/libfuse/commit/dba6b3983af34f30de01cf532dff0b66f0ed6045)]

For example, `sshfs` built against the present libfuse 3.11.0,

`$ sshfs 127.0.0.1:/home/test/testdir /home/test/sshfs_mnt -o atime`

terminates successfully (with the mount point established), while when built against 3.16.1, it outputs the error message ``fuse: unknown option(s): `-o atime'`` and terminates with exit status 1.

libfuse 3.12.0 introduced a new API version [[commit](https://github.com/libfuse/libfuse/commit/30a126c5f9009e8ff8e369c563eb941679bec252)]. From what I can tell, since file system implementations are supposed to define themselves the FUSE API version they want to be compiled against (failing that, nixpkgs would have to set it via a build parameter and keep it in sync with upstream), this is supposed to come into play on an opt-in basis only, and the change should be fully backwards-compatible to current derivations that (directly) depend on `fuse3`.

## Things done

Tested using `sshfs`.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

6t8k committed:
fuse3: 3.11.0 -> 3.16.1

https://github.com/libfuse/libfuse/blob/fuse-3.16.1/ChangeLog.rst#libfuse-3161-2023-08-08

One change can be expected to break some setups:

- Unsupported mount options are no longer silently accepted [1]

For example, sshfs built against the present libfuse 3.11.0,

`$ sshfs 127.0.0.1:/home/test/testdir /home/test/sshfs_mnt -o atime`

terminates successfully (with the mount point established), while when
built against 3.16.1, it outputs the error message `fuse: unknown
option(s): `-o atime'` and terminates with exit status 1.

Kiskae commented on pkgs/os-specific/linux/fuse/common.nix:
```diff
@@ -17,7 +17,7 @@ in stdenv.mkDerivation rec {
     repo = "libfuse";
     rev = "${pname}-${version}";
-    sha256 = sha256Hash;
+    hash = hash;
```
```suggestion
    inherit hash;
```

6t8k committed:
fuse3: 3.16.1 -> 3.16.2

https://github.com/libfuse/libfuse/blob/fuse-3.16.2/ChangeLog.rst#libfuse-3162-2023-10-10
https://github.com/libfuse/libfuse/compare/fuse-3.16.1...fuse-3.16.2

symphorien commented:
can you add release notes about the breaking change you mention?

6t8k committed:
nixos/doc: document backward incompatibility of fuse3 bump to 3.16.2

symphorien commented:
thank you for bearing with us!
rebasing is not a problem, and since the contributing guide insists on the shape of commits, cannot be avoided in many cases.

xd003 commented:
Sorry for bothering but how much time would it take for this package to land up in nixpkgs unstable ? It still shows v3.11.0 there.
